src/Entity/ListBuilder/DeveloperAppListBuilder.php

```diff
@@ -219,86 +260,119 @@ public function buildHeader() {
+    $row['data'] += parent::buildRow($developer_app);
+    $rows[$infoRowId] = $row;
+    return $rows;
+  }
+
+  /**
+   * Checks the developer app and returns information for displaying warnings.
+   *
+   * Helper function for buildWarningRow(). Information about credentials,
+   * products can be returned.
+   *
+   * @param \Drupal\apigee_edge\Entity\DeveloperAppInterface $developer_app
+   *   Developer app entity to be checked.
+   *
+   * @return array
+   *   An array containing the information about the credentials, products for
+   *   displaying warning messages.
+   */
+  protected function checkDeveloperApp(DeveloperAppInterface $developer_app): array {
+    $info = [];
+    $info['hasRevokedCred'] = FALSE;
+    $info['hasRevokedCredProduct'] = FALSE;
+    $info['hasPendingCredProduct'] = FALSE;
+    $info['problematicApiProductName'] = NULL;
+
+    foreach ($developer_app->getCredentials() as $credential) {
       if ($credential->getStatus() === AppCredential::STATUS_REVOKED) {
-        $hasRevokedCred = TRUE;
+        $info['hasRevokedCred'] = TRUE;
         break;
       }
       foreach ($credential->getApiProducts() as $credProduct) {
         if ($credProduct->getStatus() == CredentialProduct::STATUS_REVOKED) {
-          $problematicApiProductName = $credProduct->getApiproduct();
-          $hasRevokedCredProduct = TRUE;
+          $info['problematicApiProductName'] = $credProduct->getApiproduct();
+          $info['hasRevokedCredProduct'] = TRUE;
           break;
         }
         elseif ($credProduct->getStatus() == CredentialProduct::STATUS_PENDING) {
-          $problematicApiProductName = $credProduct->getApiproduct();
-          $hasPendingCredProduct = TRUE;
+          $info['problematicApiProductName'] = $credProduct->getApiproduct();
+          $info['hasPendingCredProduct'] = TRUE;
           break;
         }
       }
     }
 
-    /*
-     * Display warning sign next to the status if app's status is approved, but:
-     *  - any credentials of the app is in revoked status
-     *  - any products of any credentials of the app is in revoked or pending
-     *    status.
-     */
-    if ($entity->getStatus() === App::STATUS_APPROVED && ($hasRevokedCred || $hasPendingCredProduct || $hasRevokedCredProduct)) {
-      $build['status'] = $infoRow['app_status']['data'];
+    return $info;
+  }
+
+  /**
+   * Builds a warning row for a developer app in the entity listing.
+   *
+   * The warning row contains the warning messages if necessary.
+   *
+   * @param \Drupal\apigee_edge\Entity\DeveloperAppInterface $developer_app
+   *   The developer app entity for this row of the list.
+   * @param array $rows
+   *   A render array containing the info and warning rows to be displayed.
+   *
+   * @return array
+   *   A render array structure of fields for this developer app entity.
+   */
+  protected function buildWarningRow(DeveloperAppInterface $developer_app, array $rows): array {
```

Comment:
Why do you pass `$rows` if you only use it for one purpose, adding a new warning row to it if there were warning  and then return it?
You should either return only the warning row in this method (or an empty array for example) or pass $rows by reference and do not return anything in this method. I'd prefer the first version.